jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,460 @@
+         */
+        public boolean tryAcquire()
+        {
+            while (true)
+            {
+                long encoded = state.get();
+                int usageCount = AtomicBiInteger.getHi(encoded);
+                boolean closed = usageCount < 0;
+                int multiplexingCount = AtomicBiInteger.getLo(encoded);
+                int currentMaxUsageCount = maxUsageCount;
+                if (closed || multiplexingCount >= maxMultiplex || (currentMaxUsageCount > 0 && usageCount >= currentMaxUsageCount))
+                    return false;
+
+                if (state.compareAndSet(encoded, usageCount + 1, multiplexingCount + 1))
+                    return true;
+            }
+        }
+
+        /**
+         * @return false if the pooled object should be disposed of.
+         */
+        public boolean tryRelease()
+        {
+            int newMultiplexingCount;
+            int usageCount;
+            while (true)
+            {
+                long encoded = state.get();
+                usageCount = AtomicBiInteger.getHi(encoded);
+                boolean closed = usageCount < 0;
+                if (closed)
+                    return false;
+
+                newMultiplexingCount = AtomicBiInteger.getLo(encoded) - 1;
+                if (newMultiplexingCount < 0)
+                {
+                    if (LOGGER.isDebugEnabled())
+                        LOGGER.debug("try release failed on negative multiplexing count : {}", newMultiplexingCount);
+                    return false;
+                }
+
+                if (state.compareAndSet(encoded, usageCount, newMultiplexingCount))
+                    break;
+            }
+
+            int currentMaxUsageCount = maxUsageCount;
+            boolean overUsed = currentMaxUsageCount > 0 && usageCount >= currentMaxUsageCount;
+            return !(overUsed && newMultiplexingCount == 0);
+        }
+
+        /**
+         * @return true if the pooled object was marked as removed.
+         */
+        public boolean tryRemove()
+        {
+            boolean removed;
+            int lastUsageCount = state.getHi();
+            while (true)
+            {
+                long encoded = state.get();
+                int usageCount = AtomicBiInteger.getHi(encoded);
+                int currentMaxUsageCount = maxUsageCount;
+                if (currentMaxUsageCount > 0 && usageCount >= currentMaxUsageCount)
+                {
+                    removed = state.compareAndSetHi(usageCount, -1);
+                }
+                else
+                {
+                    removed = state.compareAndSet(usageCount, -1, 1, 0);
+                    if (!removed)
+                        removed = state.compareAndSet(usageCount, -1, 0, 0);
+                }
+                if (removed || lastUsageCount == usageCount)
+                    break;
+                lastUsageCount = state.getHi();
+            }
+            return removed;
```

Comment:
I would rewrite this in a way similar to `tryRelease()`.
We want to set `hi=-1` and return `true` only if also the `multiplexCount==0`.
I would read the `multiplexCount`, update `closed = casHi(usageCount, -1)` and then `return closed && multiplexCount==0`.
Am I missing something?

Reply:
Summary of our review: the state should be CAS'ed to hi: -1, lo: Math.max(multiplexCount - 1).